test/unit/lua_func_adapter.c

```diff
@@ -0,0 +1,267 @@
+#include "diag.h"
+#include "fiber.h"
+#include "memory.h"
+#include "lua/init.h"
+#include "lua/error.h"
+#include "lua/utils.h"
+#include "lua/msgpack.h"
+#include "box/tuple.h"
+#include "box/lua/func_adapter.h"
+#include "box/lua/tuple.h"
+#include "core/func_adapter.h"
+
+#define UNIT_TAP_COMPATIBLE 1
+#include "unit.h"
+#include "lua_test_utils.h"
+
+#define EPS 0.0001
+
+/**
+ * Check if two floating point numbers are equal.
+ */
+static bool
+number_eq(double a, double b)
+{
+	return fabs(a - b) < EPS;
+}
+
+#undef EPS
+
+static struct lua_State *L;
+static struct func_adapter_ctx ctx;
+
+/**
+ * Generates Lua function from string and returns reference to it in
+ * Lua registry.
+ */
+static int
+generate_function(const char *function)
+{
+	int rc = luaT_dostring(L, tt_sprintf("return %s", function));
+	fail_if(rc != 0);
+	return lua_gettop(L);
+}
+
+static void
+test_numeric(void)
+{
+	plan(5);
+	header();
+
+	int idx = generate_function(
+		"function(a, b, c, d) "
+		"return a * b * c * d, a + b + c + d end");
+	struct func_adapter *func = lua_func_adapter_create(L, idx);
+	func_adapter_begin(func, &ctx);
+	func_adapter_push_number(func, &ctx, 3);
+	func_adapter_push_integer(func, &ctx, 5);
+	func_adapter_push_integer(func, &ctx, 7);
+	func_adapter_push_number(func, &ctx, 11);
+	int rc = func_adapter_call(func, &ctx);
+	fail_if(rc != 0);
+	ok(func_adapter_is_number(func, &ctx), "Expected number");
+	double number_retval = 0;
+	func_adapter_pop_number(func, &ctx, &number_retval);
+	ok(func_adapter_is_integer(func, &ctx), "Expected integer");
+	int64_t integer_retval = 0;
+	func_adapter_pop_integer(func, &ctx, &integer_retval);
+	ok(func_adapter_is_null(func, &ctx), "Expected null - no values left")
+	func_adapter_end(func, &ctx);
+	func_adapter_destroy(func);
+
+	double number_expected = 3 * 5 * 7 * 11;
+	int64_t integer_expected = 3 + 5 + 7 + 11;
+	ok(number_eq(number_expected, number_retval),
+	   "Returned number must be as expected");
+	is(integer_retval, integer_expected,
+	   "Returned integer must be as expected");
+
```

Comment:
You should reset the Lua stack after each test (`lua_settop(L, 0)`) otherwise they might affect each other.

Reply:
Done.